{"cells":[{"cell_type":"markdown","metadata":{"id":"qB7JOs0EZrrY"},"source":["# **Session 4 - Sequence Alignment**"]},{"cell_type":"markdown","metadata":{"id":"mFZqUCpMZrrY"},"source":["<u>Sequence Alignment</u>\n","  + A method of arranging sequences of DNA, RNA, or protein sequences to identify regions of similarity. \n","  + The similarity being identified, may be a result of functional, structural, or evolutionary relationships between the sequences.\n","  + Identifeis similarity and homology\n","  + Homology: descent from a common ancestor or source.\n","\n","![](https://drive.google.com/uc?export=view&id=1xjcoAfhvq0JY-Oc7EiVH-syEe5qamdrw)\n","\n","+ Terms\n","  + Matches\n","  + Mismatches\n","  + Gap\n","\n","<br></br>\n","<u>Alignment Types</u>\n","+ Global alignment: finds the best concordance/agreement between all characters in two sequences\n","    + Mostly from end to end\n","    + By Needle\n","+ Local Alignment: finds just the subsequences that align the best\n","    + In this method, we consider subsequences within each of the 2 sequences and try to match them to obtain the best alignment.\n","    + By Water\n"," \n","![](https://drive.google.com/uc?export=view&id=1NRwK49u9zjKN9KjiJZyBprlYFr6PPWe5)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"CzjHokWob2Ou"},"outputs":[],"source":["# Install and import Bio\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:49:11.476142Z","start_time":"2022-09-16T06:49:11.451206Z"},"id":"9moTFXCKZrrY"},"outputs":[],"source":["# Import the required functions (pairwise2 and format_alignment) from Bio packages\n","# Import Seq class\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:49:30.754744Z","start_time":"2022-09-16T06:49:30.746774Z"},"id":"mX2RIvV1ZrrZ"},"outputs":[],"source":["# create example sequences\n","# TCACTCGT\n","# ATTCG\n","\n"]},{"cell_type":"markdown","source":["**Global Alignment**\n","\n","> When to use?\n","+ 2 Sequences are quite similar\n","+ 2 Sequences have approximately the same length\n","\n","> Examples\n","+ Emboss NEEDLE\n","+ Needleman-Wunsch Global Align"],"metadata":{"id":"NVzhugoTSU_f"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:49:31.257546Z","start_time":"2022-09-16T06:49:31.237634Z"},"id":"1GShHX88ZrrZ"},"outputs":[],"source":["# Perform Global Alignment\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:28.727885Z","start_time":"2022-09-16T06:50:28.707906Z"},"id":"bO4vD5ovZrrZ"},"outputs":[],"source":["# display the alignment\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:29.529428Z","start_time":"2022-09-16T06:50:29.503529Z"},"id":"joqq3YqPZrrZ"},"outputs":[],"source":["# View all possible alignment\n","\n"]},{"cell_type":"markdown","source":["**Local Alignment**\n","\n","> When to use?\n","+ 2 sequences have a small matched region\n","+ 2 Sequences are of different lengths\n","+ One sequences is a subsequences of the other\n","\n","> Examples\n","+ Blast\n","+ Emboss WATER\n","+ Lalign"],"metadata":{"id":"hIBqIqUoSYEt"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:30.892311Z","start_time":"2022-09-16T06:50:30.867386Z"},"id":"DPK31YekZrra"},"outputs":[],"source":["# Perform Local Alignment\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:31.584717Z","start_time":"2022-09-16T06:50:31.570751Z"},"id":"GxV_shdIZrra"},"outputs":[],"source":["# View all possible alignment\n","\n"]},{"cell_type":"markdown","source":["**Alignment Scores**"],"metadata":{"id":"AsVrKvPJcXKB"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:32.086768Z","start_time":"2022-09-16T06:50:32.069787Z"},"id":"MW8XNhLZZrra"},"outputs":[],"source":["# Get Global Alignment's score\n","\n"]},{"cell_type":"code","source":["# Get Local Alignment's score\n","\n"],"metadata":{"id":"MP2Aw_wLctR8"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"xvz5rYMRZrra"},"source":["---\n","\n","# Percentage of Similarity with Alignment\n","\n","\n"]},{"cell_type":"markdown","source":["+ Check for similarity or percentage of similarity using Alignment\n","+ fraction of nucleotides that is the same/ total number of nucleotides * 100%"],"metadata":{"id":"Gouy4tLwRZUv"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:35.123020Z","start_time":"2022-09-16T06:50:35.110056Z"},"id":"34J5HokSZrrb"},"outputs":[],"source":["# Calculate global alignment's percentage of similarity\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:50:38.951540Z","start_time":"2022-09-16T06:50:38.946552Z"},"id":"oDG-4cTNZrrb"},"outputs":[],"source":["# Calculate local alignment's percentage of similarity\n","\n"]},{"cell_type":"markdown","metadata":{"id":"r1pe80qpZrrc"},"source":["---\n","# Global Alignments with Maximum Similarity Score\n"]},{"cell_type":"markdown","source":["Find out all the possible global alignments with the maximum similarity score\n","+ Matches &ensp;&ensp;&ensp;&ensp;&ensp;&ensp; : + 2 points \n","+ Mismatches &ensp;&ensp;&ensp; : - 1 point\n","+ Opening a gap&ensp;&ensp;: - 0.5 point\n","+ Extending a gap : - 0.1 point"],"metadata":{"id":"JHFufBAjgfdG"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:53:34.399930Z","start_time":"2022-09-16T06:53:34.374965Z"},"id":"BYkNrR0UZrrc"},"outputs":[],"source":["# Perform Global alignment with maximum similarity score\n","# globalms(seqA, seqB, match, mismatch, gap, extend)\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:53:36.183412Z","start_time":"2022-09-16T06:53:36.165429Z"},"id":"Vp86oIRWZrrc"},"outputs":[],"source":["# View all possible alignment\n","\n"]},{"cell_type":"markdown","source":["---\n","# Similarity Between Sequences"],"metadata":{"id":"OvWuer1DDtlC"}},{"cell_type":"markdown","metadata":{"id":"qJhF7AatZrrd"},"source":["\n","+ Sequence Alignment\n","    - Dynamic Programming (Global/Local/(needle/water))\n","    - Dotplot\n","    \n","+ Similarity: resemblance between two sequences in comparison\n","    - the minimal number of edit operations (inserts, deletes, and substitutions) in order to transform the one sequence into an exact copy of the other sequence being aligned \n","    - distance\n","+ Identity: the number of charaters that match EXACTLY between two different sequences\n","    + Gaps are not counted \n","    + The measurement is relational to the shorter of the two sequences. \n","    + This has the effect that sequence identity is not transitive, i.e. \n","    + if sequence A=B and B=C then A is not necessarily equal C (in terms of the identity distance measure) :\n"," \n"," <br></br>\n","\n","\n","```\n"," A: AAGGCTT\n"," B: AAGGC\n"," C: AAGGCAT\n","```\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"JUcFkdiyZrrd","jupyter":{"source_hidden":true}},"source":["+ 100% identity does not mean two sequences are the same.\n","  + identity(A,B) = 100% *(5 identical nucleotides / min(length(A),length(B)))*\n","\n","  + Identity(B,C)=100% *(5 identical nucleotides / min(length(A),length(B)))*\n","  + identity(A,C)=85% *(6 identical nucleotides / 7)*\n","+ Sequence similarity is first of all a general description of a relationship but nevertheless its more or less common practice to define similarity as an optimal matching problem (for sequence alignments or unless defined otherwise). \n","+ Hereby, the optimal matching algorithm finds the minimal number of edit operations (inserts, deletes, and substitutions) in order to transform the one sequence into an exact copy of the other sequence being aligned (edit distance). \n","+ Using this, the percentage sequence similarity of the examples above are sim(A,B)=60%, sim(B,C)=60%, sim(A,C)=86% (semi-global, sim=1-(edit distance/unaligned length of the shorter sequence)). But there are other ways to define similarity between two objects (e.g. using tertiary strucure of proteins).\n","An then you might start to conclude from similarity to homology, but this was already covered sufficiently\n","+ read more https://www.researchgate.net/post/Homology_similarity_and_identity-can_anyone_help_with_these_terms"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"1wKp0HNYZrrd"},"outputs":[],"source":["# Create 3 example sequences\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"bu0tTnl3Zrre"},"outputs":[],"source":["# Calculate local alignment and get the scores\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"EZF4Y45KZrrf"},"outputs":[],"source":["# Calculate percentage of similarity and print\n","\n"]},{"cell_type":"code","source":["# Check Concept : Does 100% similarity score means the sequence is exactly the same?\n","\n"],"metadata":{"id":"ir2whBAp3ZAr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zIuwMA0IZrrf"},"source":["---\n","# Hamming distance\n","\n","Shows how many places 2 strings differ\n","\n"]},{"cell_type":"markdown","source":["+ Hamming distance between two strings of equal length is the number of positions at which the corresponding symbols are different. \n","+ In other words, it measures the minimum number of substitutions required to change one string into the other, or the minimum number of errors that could have transformed one string into the othe\n","+ It is used for error detection or error correction\n","+ It is used to quantify the similarity of DNA sequences,\n","+ For checking the edit distance\n"," - edit distance is a way of quantifying how dissimilar two strings (e.g., words) are to one another by counting the minimum number of operations required to transform one string into the other. \n"," - eg Levenshtein distance"],"metadata":{"id":"wdjZHzQWDlWG"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"IUkYPKwvZrrg"},"outputs":[],"source":["# Create example sequences\n","# ACTAT\n","# ACTTA\n","# ACTT\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"GggjZ-bBZrrg"},"outputs":[],"source":["# Define Hamming Distance function\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"oO2SG07XZrrg"},"outputs":[],"source":["# Perform Hamming Distance calculation\n","\n"]},{"cell_type":"markdown","metadata":{"id":"vL8Zb07dZrrh"},"source":["---\n","# Levenshtein Distance"]},{"cell_type":"markdown","source":["+  This method was invented in 1965 by the Russian Mathematician Vladimir Levenshtein (1935-2017).\n","+  The distance value describes the minimal number of deletions, insertions, or substitutions that are required to transform one string (the source) into another (the target).\n","+  Unlike the Hamming distance, the Levenshtein distance works on strings with an unequal length."],"metadata":{"id":"J4KJKrafDV3h"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"GnV34nMpZrrh"},"outputs":[],"source":["# Install Levenshtein Distance library\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Ndzy8-IltIbJ"},"outputs":[],"source":["# Import distance function from Levenshtein\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4pYUA-pMZrri"},"outputs":[],"source":["# Perform Levenshtein Distance calculation\n","\n"]},{"cell_type":"markdown","metadata":{"id":"9ehFpOODZrri"},"source":["---\n","# Dot Plot\n"]},{"cell_type":"markdown","source":["+ A dot plot is a graphical method that allows the comparison of two biological sequences \n","and identify regions of close similarity between them.\n","+ Simplest method - put a dot wherever\n","sequences are identical \n","+ Dot plots compare two sequences by organizing one sequence on the x-axis, and another on the y-axis, of a plot. \n","+ When the residues of both sequences match at the same location on the plot, a dot is drawn at the corresponding position\n","\n","<br></br>\n","<u>Usefulness</u>\n","\n","\n","+ Dot plots can also be used to visually inspect sequences for \n","  - direct or inverted repeats\n","  - regions with low sequence complexity.\n","  - Similar regions\n","  - Repeated sequences\n","  - Sequence rearrangements\n","  - RNA structures\n","  - Gene order\n","\n"],"metadata":{"id":"JlRU3fJvECxE"}},{"cell_type":"markdown","source":["<u>Creating dotplot utility functions</u>"],"metadata":{"id":"FQ0Yc4nNJxAR"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:24.087353Z","start_time":"2022-09-16T06:56:24.063424Z"},"id":"YHekyWbrZrri"},"outputs":[],"source":["# Define Delta function\n","# ---------------------\n","# Takes two arguments, x and y, and returns 0 if they are equal, and 1 otherwise.\n","# This function is used in other functions to tell whether a pair of nucleotides/amino acids matches\n","\n"]},{"cell_type":"code","source":["# Define M function\n","# -----------------\n","# A utility function that calculates the number of matches between two subsequences of the two input sequences.\n","# \n","# Param:\n","# seq1: a string representing the first sequence\n","# seq2: a string representing the second sequence\n","# i: an integer representing the starting index of the subsequence in seq1\n","# j: an integer representing the starting index of the subsequence in seq2\n","# k: an integer representing the length of the subsequences to be compared\n","\n"],"metadata":{"id":"X4kHslFeJ0Hk"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Define makeMatrix function\n","# --------------------------\n","# The function returns a matrix of match scores between all pairs of subsequences of the two input sequences\n","\n","# Param:\n","# seq1: a string representing the first sequence\n","# seq2: a string representing the second sequence\n","# k: an integer representing the length of the subsequences to be compared\n","\n"],"metadata":{"id":"ktWBU-VVJ2fy"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["<u>Creating and displaying dotplot using Matplotlib</u>"],"metadata":{"id":"Ar-c8trPPBl9"}},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:26.176446Z","start_time":"2022-09-16T06:56:26.157532Z"},"id":"Go8ZV6S6Zrrj"},"outputs":[],"source":["# Create 2 DNA sequence examples\n","# \"ACCTGAGCTCACCTGAGTTA\"\n","# \"ACCTGAGCTCACCTGAGTTA\"\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:30.055412Z","start_time":"2022-09-16T06:56:30.037423Z"},"id":"788PZTM9Zrrk"},"outputs":[],"source":["# Import Numpy & Matplotlib libraries to calculate and plot the result\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:31.186548Z","start_time":"2022-09-16T06:56:30.706270Z"},"id":"_59o6rbFZrrk"},"outputs":[],"source":["# Create a function to display dotplot using Matplotlib\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:48.256115Z","start_time":"2022-09-16T06:56:48.242149Z"},"id":"jE2FA1etZrrk"},"outputs":[],"source":["# Call the function\n","\n"]},{"cell_type":"markdown","source":["<u>Creating and displaying dotplot using ASCII characters</u>"],"metadata":{"id":"Q39NFTIePP8O"}},{"cell_type":"code","source":["# Define plotMatrix function\n","# --------------------------\n","# This function print out the matrix of match scores in a user-friendly way.\n","\n","# Param:\n","# Mat: a two-dimensional list representing the matrix of match scores\n","# t: an integer representing the threshold for determining whether a match score should be considered significant\n","# seq1: a string representing the first sequence\n","# seq2: a string representing the second sequence\n","# nonblank: a character that is used to represent significant match scores\n","# blank: a character that is used to represent insignificant match scores\n","\n"],"metadata":{"id":"mJ97BuOlJ4Vl"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Define dotplot Function\n","# -----------------------\n","# The dotplot function creates a dot plot for two sequences.\n","\n","# Param:\n","# seq1: a string representing the first sequence\n","# seq2: a string representing the second sequence\n","# k: an integer representing the length of the k-mer used to compare the sequences (default value is 1)\n","# t: an integer representing the threshold for determining whether a match score should be considered significant (default value is 1)\n","# The function first creates a match score matrix M by calling the makeMatrix function, passing in the two sequences and the k-mer length k.\n","\n"],"metadata":{"id":"M_Sw_nmVO3K_"},"execution_count":null,"outputs":[]},{"cell_type":"code","execution_count":null,"metadata":{"ExecuteTime":{"end_time":"2022-09-16T06:56:27.263241Z","start_time":"2022-09-16T06:56:27.247285Z"},"id":"3vedlnj2Zrrj"},"outputs":[],"source":["# Run Dot Plot function\n","\n"]},{"cell_type":"markdown","source":["*Read more : https://stackoverflow.com/questions/40822400/how-to-create-a-dotplot-of-two-dna-sequence-in-python*"],"metadata":{"id":"RnnhFsS9Gfnb"}}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.4"},"toc":{"base_numbering":1,"nav_menu":{},"number_sections":true,"sideBar":true,"skip_h1_title":false,"title_cell":"Table of Contents","title_sidebar":"Contents","toc_cell":false,"toc_position":{},"toc_section_display":true,"toc_window_display":false}},"nbformat":4,"nbformat_minor":0}